### PR TITLE
Fix getHomeDirectory on POSIX systems. Fixes #102

### LINKED ISSUE #102: getHomeDirectory on POSIX systems is unreliable
https://github.com/haskell/directory/blob/b3184cabb8c6a90eec7e921d6ca5058d57da3d82/System/Directory/Internal/Posix.hsc#L283-L284

strikes me as a naive implementation. [POSIX](https://pubs.opengroup.org/onlinepubs/009695399/basedefs/xbd_chap08.html) doesn't say that `$HOME` is mandatory to be set. The wording is:

> If the variables in the following two sections are present in the environment during the execution of an application or utility, they shall be given the meaning described below.

Even worse, `getHomeDirectory` will **error** when this optional environment variable is not set.

A better way of doing this would be:

* if $HOME is set in the environment, return it
* if $HOME is not set in the environment, look up `getpwuid(getuid())->pw_dir`, see [getpwuid](https://pubs.opengroup.org/onlinepubs/009695399/functions/getpwuid.html) (or rather `getpwuid_r` for thread-safety)
* if none are set, raise an error

Inspecting other popular standard library language functions, they seem to do exactly that:

* https://doc.rust-lang.org/std/env/fn.home_dir.html
* https://docs.python.org/3/library/os.path.html#os.path.expanduser

The only thing that is debatable about this is how an *empty* and *set* `$HOME` will be treated.

### PR DESCRIPTION
**Pull request by hasufell**

### CODE REVIEW

**Comment by hasufell** [comment]
This changes semantics in a possibly non-backwards compatible manner. If people have previously relied on this function to error out when `$HOME` is not set, then this patch will break their code. As such, this justifies a version bump to `1.4.0.0`.

File: System/Directory/Internal/Posix.hsc [hunk 1]
```diff
@@ -280,8 +281,14 @@ setFileTimes' pth atime' mtime' =
...
 
+-- | $HOME is preferred, because the user has control over it. However, POSIX
+-- doesn't define it as a mandatory variable, so fall back to `getpwuid_r`.
 getHomeDirectoryInternal :: IO FilePath
-getHomeDirectoryInternal = getEnv "HOME"
+getHomeDirectoryInternal = do
+  e <- lookupEnv "HOME"
+  case e of
+       Just fp -> pure fp
+       _ -> fmap PU.homeDirectory $ PU.getEffectiveUserID >>= PU.getUserEntryForID
```
Comment (Rufflewind) [inline System/Directory/Internal/Posix.hsc]: Consider writing `Nothing ->` explicitly to make the intent clear.

Comment (Rufflewind) [reply System/Directory/Internal/Posix.hsc]: And please use `<$>` to be consistent with the current style.

File: System/Directory.hs [hunk 2]
```diff
@@ -1498,9 +1498,12 @@ but note that it isn't generally considered good practice to store
 application-specific data here; use 'getXdgDirectory' or
 'getAppUserDataDirectory' instead.
 
-On Unix, 'getHomeDirectory' returns the value of the @HOME@
-environment variable.  On Windows, the system is queried for a
-suitable path; a typical path might be @C:\/Users\//\<user\>/@.
+On Unix, 'getHomeDirectory' behaves as follows:
+
+* Returns $HOME env variable if set (including to an empty string).
+* Otherwise uses home directory returned by `getpwuid_r` using the UID of the current proccesses user. This basically reads the /etc/passwd file. An empty home directory field is considered valid.
```
Comment (Rufflewind) [inline System/Directory.hs]: Can we add a test case for this?  Unset `HOME` and then try calling this function to make sure it does not fail.

**Comment by hasufell** [comment]
> I understand the concern but this seems like a very unusual edge case that should not be relied upon since: (1) It concerns the failure scenario; (2) It relies on a particular nuance of the APIs platform-dependent behavior.

Sure, your choice.

File: tests/GetHomeDirectory002.hs [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+-- Test that the getpwuid_r fallback works.
+-- This is only relevant on unix.
+main :: TestEnv -> IO ()
+main _t = do
+#if !defined(mingw32_HOST_OS)
+  unsetEnv "HOME"
+#endif
+  _ <- getHomeDirectory
+  T(expect) () True -- avoid warnings about redundant imports
+  return ()
```
Comment (Rufflewind) [inline tests/GetHomeDirectory002.hs]: Is this line needed?

File: tests/GetHomeDirectory002.hs [hunk 4]
```diff
@@ -0,0 +1,18 @@
+{-# LANGUAGE CPP #-}
+module GetHomeDirectory002 where
+
+#if !defined(mingw32_HOST_OS)
+import System.Posix.Env
+#endif
+#include "util.inl"
+
+-- Test that the getpwuid_r fallback works.
+-- This is only relevant on unix.
```
Comment (Rufflewind) [inline tests/GetHomeDirectory002.hs]: Does this need to be Unix-specific?  I imagine it should succeed on Windows too.

Comment (hasufell) [reply tests/GetHomeDirectory002.hs]: Windows does not have `getpwuid_r`.

Comment (Rufflewind) [reply tests/GetHomeDirectory002.hs]: The test should still work on Windows since `getHomeDirectory` has a different implementation on Windows that does not rely on `HOME`.  Thus, there is no reason to complicate this test with `#if`s.

Comment (hasufell) [reply tests/GetHomeDirectory002.hs]: That does not work. `unsetEnv` does not exist on windows.

Comment (Rufflewind) [reply tests/GetHomeDirectory002.hs]: Oh, I see.  You are using `unsetEnv` from `unix` not `base`.  In that case, let's leave this be.
